src/check/canonicalize.zig

```diff
@@ -78,6 +82,23 @@ const CalledVia = base.CalledVia;
 
 const TypeVar = types.Var;
 const Content = types.Content;
+
+/// Context for tracking lambda captures during canonicalization
+const FunctionContext = struct {
+    depth: u32,
+    captures: std.ArrayList(CIR.Expr.Capture),
```

Comment:
Rather than an ArrayList of ArrayLists, I'd suggest two array lists:
1. capture_stack: ArrayList(CIR.Expr.Capture)
2. capture_stack_frames: ArrayList(u32) 

The pattern would be: every time you enter a function, you do `capture_stack_frames.push(capture_stack.len)` and then on leaving the frame you do `capture_stack_frames.pop()`, possibly with an assert that you got back to where you expected to be. This way you can avoid a bunch of allocations.

Reply:
Let's do this in a follow up -- I'd like to merge this so I can rebase my redo-functions PR

Reply:
Resolved in https://github.com/roc-lang/roc/pull/8101